sweep/sweeper.go

```diff
@@ -1547,8 +1551,21 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	// transactions like funding of a channel.
 	sweepWithLock := func(set InputSet) error {
 		return s.cfg.Wallet.WithCoinSelectLock(func() error {
+			extraBudgetRes := fn.MapOptionZ(
```

Comment:
```diff
diff --git a/server.go b/server.go
index b99e9be44..e97917be2 100644
--- a/server.go
+++ b/server.go
@@ -1125,7 +1125,7 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	}
 
 	aggregator := sweep.NewBudgetAggregator(
-		cc.FeeEstimator, sweep.DefaultMaxInputsPerTx,
+		cc.FeeEstimator, sweep.DefaultMaxInputsPerTx, implCfg.AuxSweeper,
 	)
 
 	s.txPublisher = sweep.NewTxPublisher(sweep.TxPublisherConfig{
@@ -1151,7 +1151,6 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 		Aggregator:           aggregator,
 		Publisher:            s.txPublisher,
 		NoDeadlineConfTarget: cfg.Sweeper.NoDeadlineConfTarget,
-		AuxSweeper:           implCfg.AuxSweeper,
 	})
 
 	s.utxoNursery = contractcourt.NewUtxoNursery(&contractcourt.NurseryConfig{
diff --git a/sweep/aggregator.go b/sweep/aggregator.go
index 6028adca3..71c9baa97 100644
--- a/sweep/aggregator.go
+++ b/sweep/aggregator.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/btcsuite/btcd/btcutil"
 	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
 	"github.com/lightningnetwork/lnd/input"
 	"github.com/lightningnetwork/lnd/lntypes"
 	"github.com/lightningnetwork/lnd/lnwallet"
@@ -31,6 +32,10 @@ type BudgetAggregator struct {
 	// maxInputs specifies the maximum number of inputs allowed in a single
 	// sweep tx.
 	maxInputs uint32
+
+	// AuxSweeper is an optional interface that can be used to modify the
+	// way sweep transaction are generated.
+	auxSweeper fn.Option[AuxSweeper]
 }
 
 // Compile-time constraint to ensure BudgetAggregator implements UtxoAggregator.
@@ -38,11 +43,12 @@ var _ UtxoAggregator = (*BudgetAggregator)(nil)
 
 // NewBudgetAggregator creates a new instance of a BudgetAggregator.
 func NewBudgetAggregator(estimator chainfee.Estimator,
-	maxInputs uint32) *BudgetAggregator {
+	maxInputs uint32, auxSweeper fn.Option[AuxSweeper]) *BudgetAggregator {
 
 	return &BudgetAggregator{
-		estimator: estimator,
-		maxInputs: maxInputs,
+		estimator:  estimator,
+		maxInputs:  maxInputs,
+		auxSweeper: auxSweeper,
 	}
 }
 
@@ -159,7 +165,7 @@ func (b *BudgetAggregator) createInputSets(inputs []SweeperInput,
 
 		// Create an InputSet using the max allowed number of inputs.
 		set, err := NewBudgetInputSet(
-			currentInputs, deadlineHeight,
+			currentInputs, deadlineHeight, b.auxSweeper,
 		)
 		if err != nil {
 			log.Errorf("unable to create input set: %v", err)
@@ -173,7 +179,7 @@ func (b *BudgetAggregator) createInputSets(inputs []SweeperInput,
 	// Create an InputSet from the remaining inputs.
 	if len(remainingInputs) > 0 {
 		set, err := NewBudgetInputSet(
-			remainingInputs, deadlineHeight,
+			remainingInputs, deadlineHeight, b.auxSweeper,
 		)
 		if err != nil {
 			log.Errorf("unable to create input set: %v", err)
diff --git a/sweep/mock_test.go b/sweep/mock_test.go
index f999a4aa0..cd5db358a 100644
--- a/sweep/mock_test.go
+++ b/sweep/mock_test.go
@@ -233,9 +233,7 @@ func (m *MockInputSet) FeeRate() chainfee.SatPerKWeight {
 // AddWalletInputs adds wallet inputs to the set until a non-dust
 // change output can be made. Return an error if there are not enough
 // wallet inputs.
-func (m *MockInputSet) AddWalletInputs(wallet Wallet,
-	_ btcutil.Amount) error {
-
+func (m *MockInputSet) AddWalletInputs(wallet Wallet) error {
 	args := m.Called(wallet)
 
 	return args.Error(0)
@@ -243,7 +241,7 @@ func (m *MockInputSet) AddWalletInputs(wallet Wallet,
 
 // NeedWalletInput returns true if the input set needs more wallet
 // inputs.
-func (m *MockInputSet) NeedWalletInput(_ btcutil.Amount) bool {
+func (m *MockInputSet) NeedWalletInput() bool {
 	args := m.Called()
 
 	return args.Bool(0)
diff --git a/sweep/sweeper.go b/sweep/sweeper.go
index 0b80fe904..6257faac1 100644
--- a/sweep/sweeper.go
+++ b/sweep/sweeper.go
@@ -362,10 +362,6 @@ type UtxoSweeperConfig struct {
 	// NoDeadlineConfTarget is the conf target to use when sweeping
 	// non-time-sensitive outputs.
 	NoDeadlineConfTarget uint32
-
-	// AuxSweeper is an optional interface that can be used to modify the
-	// way sweep transaction are generated.
-	AuxSweeper fn.Option[AuxSweeper]
 }
 
 // Result is the struct that is pushed through the result channel. Callers can
@@ -1551,21 +1547,8 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	// transactions like funding of a channel.
 	sweepWithLock := func(set InputSet) error {
 		return s.cfg.Wallet.WithCoinSelectLock(func() error {
-			extraBudgetRes := fn.MapOptionZ(
-				s.cfg.AuxSweeper,
-				func(s AuxSweeper) fn.Result[btcutil.Amount] {
-					return s.ExtraBudgetForInputs(
-						set.Inputs(),
-					)
-				},
-			)
-			extraBudget, err := extraBudgetRes.Unpack()
-			if err != nil {
-				return err
-			}
-
 			// Try to add inputs from our wallet.
-			err = set.AddWalletInputs(s.cfg.Wallet, extraBudget)
+			err := set.AddWalletInputs(s.cfg.Wallet)
 			if err != nil {
 				return err
 			}
@@ -1581,21 +1564,8 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	}
 
 	for _, set := range sets {
-		extraBudgetRes := fn.MapOptionZ(
-			s.cfg.AuxSweeper,
-			func(s AuxSweeper) fn.Result[btcutil.Amount] {
-				return s.ExtraBudgetForInputs(
-					set.Inputs(),
-				)
-			},
-		)
-		extraBudget, err := extraBudgetRes.Unpack()
-		if err != nil {
-			log.Errorf("Failed to get extra budget for set %v: %v",
-				set, err)
-		}
-
-		if set.NeedWalletInput(extraBudget) {
+		var err error
+		if set.NeedWalletInput() {
 			// Sweep the set of inputs that need the wallet inputs.
 			err = sweepWithLock(set)
 		} else {
diff --git a/sweep/tx_input_set.go b/sweep/tx_input_set.go
index d26d3d75e..3777a8a37 100644
--- a/sweep/tx_input_set.go
+++ b/sweep/tx_input_set.go
@@ -37,11 +37,11 @@ type InputSet interface {
 	// AddWalletInputs adds wallet inputs to the set until a non-dust
 	// change output can be made. Return an error if there are not enough
 	// wallet inputs.
-	AddWalletInputs(wallet Wallet, extraBudget btcutil.Amount) error
+	AddWalletInputs(wallet Wallet) error
 
 	// NeedWalletInput returns true if the input set needs more wallet
 	// inputs.
-	NeedWalletInput(extraBudget btcutil.Amount) bool
+	NeedWalletInput() bool
 
 	// DeadlineHeight returns an absolute block height to express the
 	// time-sensitivity of the input set. The outputs from a force close tx
@@ -111,6 +111,12 @@ type BudgetInputSet struct {
 	// deadlineHeight is the height which the inputs in this set must be
 	// confirmed by.
 	deadlineHeight int32
+
+	// extraBudget is a value that should be allocated to sweep the given
+	// set of inputs. This can be used to add extra funds to the sweep
+	// transaction, for example to cover fees for additional outputs of
+	// custom channels.
+	extraBudget btcutil.Amount
 }
 
 // Compile-time constraint to ensure budgetInputSet implements InputSet.
@@ -167,8 +173,8 @@ func validateInputs(inputs []SweeperInput, deadlineHeight int32) error {
 }
 
 // NewBudgetInputSet creates a new BudgetInputSet.
-func NewBudgetInputSet(inputs []SweeperInput,
-	deadlineHeight int32) (*BudgetInputSet, error) {
+func NewBudgetInputSet(inputs []SweeperInput, deadlineHeight int32,
+	auxSweeper fn.Option[AuxSweeper]) (*BudgetInputSet, error) {
 
 	// Validate the supplied inputs.
 	if err := validateInputs(inputs, deadlineHeight); err != nil {
@@ -186,9 +192,31 @@ func NewBudgetInputSet(inputs []SweeperInput,
 
 	log.Tracef("Created %v", bi.String())
 
+	// Optionally attach an extra budget to the input set.
+	auxSweeper.WhenSome(func(aux AuxSweeper) {
+		bi.attachExtraBudget(aux)
+	})
+
 	return bi, nil
 }
 
+// attachExtraBudget...
+func (b *BudgetInputSet) attachExtraBudget(s AuxSweeper) error {
+	extraBudgetRes := s.ExtraBudgetForInputs(b.Inputs())
+
+	extraBudget, err := extraBudgetRes.Unpack()
+	if err != nil {
+		log.Errorf("Failed to get extra budget for set %v: %v",
+			b.Inputs(), err)
+
+		return err
+	}
+
+	b.extraBudget = extraBudget
+
+	return nil
+}
+
 // String returns a human-readable description of the input set.
 func (b *BudgetInputSet) String() string {
 	inputsDesc := ""
@@ -209,13 +237,13 @@ func (b *BudgetInputSet) addInput(input SweeperInput) {
 //
 // A set may need wallet inputs when it has a required output or its total
 // value cannot cover its total budget.
-func (b *BudgetInputSet) NeedWalletInput(extraBudget btcutil.Amount) bool {
+func (b *BudgetInputSet) NeedWalletInput() bool {
 	var (
 		// budgetNeeded is the amount that needs to be covered from
 		// other inputs. We start at the value of the extra budget,
 		// which might be needed for custom channels that add extra
 		// outputs.
-		budgetNeeded = extraBudget
+		budgetNeeded = b.extraBudget
 
 		// budgetBorrowable is the amount that can be borrowed from
 		// other inputs.
@@ -276,9 +304,7 @@ func (b *BudgetInputSet) copyInputs() []*SweeperInput {
 // set to its initial state by removing any wallet inputs added.
 //
 // NOTE: must be called with the wallet lock held via `WithCoinSelectLock`.
-func (b *BudgetInputSet) AddWalletInputs(wallet Wallet,
-	extraBudget btcutil.Amount) error {
-
+func (b *BudgetInputSet) AddWalletInputs(wallet Wallet) error {
 	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
 	// problems around RBF rules for unconfirmed inputs. This currently
 	// ignores the configured coin selection strategy.
@@ -322,7 +348,7 @@ func (b *BudgetInputSet) AddWalletInputs(wallet Wallet,
 			pi.OutPoint(), utxo.Value)
 
 		// Return if we've reached the minimum output amount.
-		if !b.NeedWalletInput(extraBudget) {
+		if !b.NeedWalletInput() {
 			return nil
 		}
 	}
```

Reply:
Ah, much better! Thank you very much :pray: